File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionIf.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;

public class TransformerFunctionIf extends TransformerFunction {
    public TransformerFunctionIf() {
        super(FunctionDescription.of(
            Map.of(
            "then", ArgumentType.of(ArgType.Any).position(0).defaultIsNull(true),
            "else", ArgumentType.of(ArgType.Any).position(1).defaultIsNull(true)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        boolean condition;
        var adapter = context.getAdapter();
        if (context.has("then")) {
            var value = context.getJsonElement(null);
            if (adapter.isTruthy(value)) {
                return context.getJsonElement("then", true);
            } else if (context.has("else")) {
                return context.getJsonElement("else", true);
            }
        } else {
            var arr = context.getJsonArray(null);
            if (arr == null || adapter.size(arr) < 2)
                return null;
            var cje = adapter.get(arr, 0);
            if (adapter.isNull(cje)) {
                condition = false;
            } else if (adapter.isJsonBoolean(cje)) {
                condition = adapter.getBoolean(cje);
            } else {
                condition = adapter.isTruthy(context.transform(context.getPathFor(0), cje));
            }

            if (condition) {
                return context.transform(context.getPathFor(1), adapter.get(arr, 1));
            } else if (adapter.size(arr) > 2) {
                return context.transform(context.getPathFor(2), adapter.get(arr, 2));
            }
        }
        return null; // default falsy value
    }
}

File: javascript/json-transform/src/functions/TransformerFunctionIf.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import { isNullOrUndefined, isTruthy } from "../JsonHelpers";

class TransformerFunctionIf extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        then: { type: ArgType.Any, position: 0, defaultIsNull: true },
        else: { type: ArgType.Any, position: 1, defaultIsNull: true },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    let condition: boolean;
    if (context.has("then")) {
      const value = await context.getJsonElement(null);
      if (isTruthy(value)) {
        return await context.getJsonElement("then", true);
      } else if (context.has("else")) {
        return await context.getJsonElement("else", true);
      }
    } else {
      const arr = await context.getJsonArray(null);
      if (arr == null || arr.length < 2) return null;
      const cje = arr[0];
      if (isNullOrUndefined(cje)) {
        condition = false;
      } else if (typeof cje === "boolean") {
        condition = cje;
      } else {
        condition = isTruthy(await context.transform(context.getPathFor(0), cje));
      }

      if (condition) {
        return await context.transform(context.getPathFor(1), arr[1]);
      } else if (arr.length > 2) {
        return await context.transform(context.getPathFor(2), arr[2]);
      }
    }
    return null; // default falsy value
  }
}

export default TransformerFunctionIf;
